File: Baekjoon_Online_Judge/NO_02257/2257.py
# -*- encoding: utf-8 -*-
import sys
r_input = sys.stdin.readline

# -------------------------------------- #
# | Created by happiness96             | #
# | Year 2020                          | #
# | Month 08                           | #
# | Day 11                             | #
# | 2257 화학식량                      | #
# -------------------------------------- #


if __name__ == "__main__":
    chemical_formula = r_input().rstrip()
    amount_of_atom = {'H': 1, 'C': 12, 'O': 16}

    stack = []
    acc = 0

    for ch in chemical_formula:
        if ch == '(':
            stack.append(ch)

        elif ch == ')':
            while stack[-1] != '(':
                acc += stack.pop()
            
            stack.pop()
            stack.append(acc)
            acc = 0
        
        elif ch in '23456789':
            stack.append(stack.pop() * int(ch))
        
        else:
            stack.append(amount_of_atom[ch])
    
    print(sum(stack))

File: Baekjoon_Online_Judge/NO_02257/2257.cpp
#include <iostream>
using namespace std;
/*
 -------------------------------------- 
 | Created by happiness96             | 
 | Year 2020                          | 
 | Month 08                           | 
 | Day 11                             | 
 | 2257 화학식량                      | 
 -------------------------------------- 
*/

string formula;
int length, index;

int recursion(){
    int acc = 0, pre = 0;

    while(index < length){
        char val = formula[index];

        if(val == '('){
            index ++;
            pre = recursion();
            acc += pre;
        }

        else if(val == ')'){
            index ++;
            return acc;
        }

        else if(val == 'H'){
            acc ++;
            pre = 1;
            index ++;
        }

        else if(val == 'C'){
            acc += 12;
            pre = 12;
            index ++;
        }
        
        else if(val == 'O'){
            acc += 16;
            pre = 16;
            index ++;
        }
        
        else{
            acc += pre * ((int)val - '0' - 1);
            index ++;
        }
    }

    return acc;
}

int main(){
    cin >> formula;

    length = formula.length();
    index = 0;

    cout << recursion();
    return 0;
}
